src/pointer/ptr.rs

```diff
@@ -140,12 +140,40 @@ pub mod invariant {
         type Aliasing: Aliasing;
         type Alignment: Alignment;
         type Validity: Validity;
+
+        /// Invariants identical to `Self` except with a different aliasing
+        /// invariant.
+        type WithAliasing<A: Aliasing>: Invariants<
+            Aliasing = A,
+            Alignment = Self::Alignment,
+            Validity = Self::Validity,
+        >;
+
+        /// Invariants identical to `Self` except with a different alignment
+        /// invariant.
+        type WithAlignment<A: Alignment>: Invariants<
+            Aliasing = Self::Aliasing,
+            Alignment = A,
+            Validity = Self::Validity,
+        >;
+
+        /// Invariants identical to `Self` except with a different validity
+        /// invariant.
+        type WithValidity<V: Validity>: Invariants<
+            Aliasing = Self::Aliasing,
+            Alignment = Self::Alignment,
+            Validity = V,
+        >;
```

Comment:
Should we hold off on this insofar that it might make backporting subsequent improvements to `Ptr` more difficult? It seems like there's a lot of cleanup to be done that doesn't involve GATs, and we'd benefit from having `Ptr` be as similar as possible between 0.8 and 0.9.

Reply:
Discussed offline - decided to stick with this.